### PR TITLE
Add sessions

### PR DESCRIPTION
**Pull request by reneos**
This PR enables the use of sessions for Magento (using [komoju-php](https://github.com/degica/komoju-php)), replacing the use of the old legacy hosted page. This also means using the API to dynamically grab the payment methods the merchant has enabled on Komoju.

## Testing
1. Build Magento following Tim's instructions and setup the plugin with merchant secret key, etc. (🙇🏻‍♀️ ): [here](https://github.com/degica/komoju-magento/blob/master/docs/dev_setup.md)
2. Test some operations:
- creating an order and completing the komoju session. Ensure the order is then marked as "processing" on Magento admin dashboard
- Make note of the current stock level. Create an order and "cancel" the komoju session (x button). Once it redirects, ensure the original order is then "cancelled" on the Magento admin dashboard and that the stock has returned to the previous level. See that the selected items have been returned to the cart
- Create an order and select Konbini payment method. Cancel it on the komoju dashboard (make sure the webhooks are sent to your Magento endpoint) and verify that the stock for the product has been returned to previous levels and that the order is marked as canceled on the magento admin dashboard
- Go back to the Magento dashboard and update the merchant secret key so that it's wrong. Try placing an order and when you reach the payment page, verify that you see this error:
image
(It disappears quite quickly -- that's a native Magento thing AFAIK).
- Any other test cases you can think of

Some notes:
- When a payment method is selected but and someone clicks "place order" but there's some issue connecting to Komoju, the current behaviour cancels the order on Magento, returns the stock, returns the items to the customer's cart, and redirects back to the payment page. I couldn't get the usual way of showing error messages to work from the controller redirect so right now there's no error message.
- ~~Still need to translate the error message from the screenshot above~~

### CODE REVIEW

**Comment by degicat** [comment]
@jdgc please review this

**Comment by degicat** [comment]
@thoiberg please review this

File: src/app/code/Komoju/Payments/Plugin/WebhookEventProcessor.php [hunk 1]
```diff
@@ -97,18 +97,17 @@ public function processEvent()
 
             $this->order->save();
         } elseif ($this->webhookEvent->eventType() == 'payment.cancelled') {
-            $this->order->setState(Order::STATE_CANCELED);
-            $this->order->setStatus(Order::STATE_CANCELED);
-            
+            $this->order->cancel();
```
Comment (thoiberg) [inline src/app/code/Komoju/Payments/Plugin/WebhookEventProcessor.php:100]: I'm glad you found a cleaner way to do this, have to set the state _and_ the status to the same thing definitely looks confusing (and I wrote the code! 😛 )

**Comment by thoiberg** [comment]
I noticed that it nests the fields under each individual payment method. Since our payment methods have the all the same fields could we have those just display about/below the payment method fields. If this is just how Magento lays them out then it's fine to leave it as is 😄 

image

**Comment by reneos** [comment]
@thoiberg Yeah, the billing forms are controlled by Magento so they're the same for every payment method. I wanted it under each individual method because that's how the built in method works (if you click check/money order it's the same thing). If I put it under all the komoju payment methods, the "place order" button is way at the bottom and it's sort of confusing.

**Comment by thoiberg** [comment]
Not one for this PR but do you think it would look better if we nested the Payment methods under a KOMOJU header or something? I'm just thinking about the use case where the merchant has other payment methods from Paypal, etc activated.

image

**Review by thoiberg — COMMENTED** [review]
LGTM 🏆 

great job!

**Review by thoiberg — APPROVED** [review]
Didn't tick the Approve button 😅

**Comment by jdgc** [comment]
> Not one for this PR but do you think it would look better if we nested the Payment methods under a KOMOJU header or something? I'm just thinking about the use case where the merchant has other payment methods from Paypal, etc activated.
> 
> image

I think this way is good since to the end customer KOMOJU doesn't really mean much (we actually had a customer support request on woocommerce to move payment methods out of the komoju subheader in the past) - but we do probably want to avoid the case of payment methods clashing with other providers. Maybe we can add some indicator to the name of the method? 'powered by KOMOJU' or similar? Something to think about

**Comment by reneos** [comment]
> > Not one for this PR but do you think it would look better if we nested the Payment methods under a KOMOJU header or something? I'm just thinking about the use case where the merchant has other payment methods from Paypal, etc activated.
> > image
> 
> I think this way is good since to the end customer KOMOJU doesn't really mean much (we actually had a customer support request on woocommerce to move payment methods out of the komoju subheader in the past) - but we do probably want to avoid the case of payment methods clashing with other providers. Maybe we can add some indicator to the name of the method? 'powered by KOMOJU' or similar? Something to think about

Maybe a good way in the future is have some header above all the payment methods that's hidden by default and can be shown via some admin flag on the dashboard

**Comment by reneos** [comment]
@thoiberg Hey Tim, added in the toggle ([here](https://github.com/degica/komoju-magento/pull/9/commits/3e3bdf495fbe0982cf841826ffa157750928358c)) for showing the title above the payment methods. It's not displayed by default.
Here's the config page:
image
(The label text and comment text are now different from that screenshot)
And here's what it looks like with the title above the payment methods:
image
I don't think you need to test it again, but could you take a look at the diff? Thanks! 🙏🏻

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv [hunk 2]
```diff
@@ -22,7 +22,9 @@ KOMOJU,KOMOJU
...
 "The webhook URL will be at /komoju/hostedpage/webhook of this site","webhook URLは、このサイトの/ komoju / hostedpage / webhookにあります"
 "Payment Methods","お支払い方法"
-"Allow credit card payments","クレジットカード決済を許可する"
-"Allow konbini payments","コンビニ支払いを許可する"
 "Checkout Settings","決済設定"
 Title,題名
+"This plugin can only work if the store's base currency is one of: JPY","This plugin can only work if the store's base currency is one of: JPY"
+"Encountered an issue communicating with KOMOJU. Please wait a moment and try again.","KOMOJUを読み込みませんでした。しばらく待ってから再実行してください。"
+"Show Title at Checkout","チェックアウトでタイトルを表示する"
+"Display the title above the payment methods on the checkout page","チェックアウトページの支払い方法の上にタイトルを表示する"
```
Comment (reneos) [inline src/app/code/Komoju/Payments/i18n/ja_JP.csv]: @kenta-s @Yuuki77 Would one of you mind checking on these translations? 🙇🏻‍♀️ I also wasn't sure how to translate "This plugin can only work if the store's base currency is one of: JPY"

**Review by jdgc — APPROVED** [review]
Checked it out on local setup, works great

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv (hunk 2 above)
Comment (Yuuki77) [reply src/app/code/Komoju/Payments/i18n/ja_JP.csv]: @reneos 
These are a bit difficult to translate...
 Can you give me a screenshot of the page where these translations are displayed?  🙏 

base currency...

Comment (reneos) [reply src/app/code/Komoju/Payments/i18n/ja_JP.csv]: @Yuuki77 Sorry, I understand it's difficult without context!! 

I think for Magento, `base currency` is like the main currency of the store. Unfortunately I don't think Magento has official docs in japanese.. maybe we can say something like 基本通貨？
It's used under the dropdown here (whether to enable the plugin):
image

The "encountered an issue" error appears on the checkout page when the API cannot reach Komoju to retrieve the merchant's enabled payment methods:
image

The `Show Title` and `Display the title [...]` messages appear in the plugin settings. It allows the store owners to choose whether or not to show an optional header title above the payment methods on the checkout page.
image

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv [hunk 3]
```diff
@@ -22,7 +22,9 @@ KOMOJU,KOMOJU
 "Webhook Secret","Webhookシークレット"
 "The webhook URL will be at /komoju/hostedpage/webhook of this site","webhook URLは、このサイトの/ komoju / hostedpage / webhookにあります"
 "Payment Methods","お支払い方法"
-"Allow credit card payments","クレジットカード決済を許可する"
-"Allow konbini payments","コンビニ支払いを許可する"
 "Checkout Settings","決済設定"
 Title,題名
+"This plugin can only work if the store's base currency is one of: JPY","This plugin can only work if the store's base currency is one of: JPY"
+"Encountered an issue communicating with KOMOJU. Please wait a moment and try again.","KOMOJUを読み込みませんでした。しばらく待ってから再実行してください。"
+"Show Title at Checkout","チェックアウトでタイトルを表示する"
```
Comment (Yuuki77) [inline src/app/code/Komoju/Payments/i18n/ja_JP.csv]:
```suggestion
"Show Title at Checkout","チェックアウトでタイトルを表示"
```

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv (hunk 2 above)
Comment (Yuuki77) [inline src/app/code/Komoju/Payments/i18n/ja_JP.csv]: I think original text is fine but I think you do not need する here 
```suggestion
"Display the title above the payment methods on the checkout page","チェックアウトページの支払い方法の上にタイトルを表示"
```

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv [hunk 4]
```diff
@@ -22,7 +22,9 @@ KOMOJU,KOMOJU
 "Webhook Secret","Webhookシークレット"
 "The webhook URL will be at /komoju/hostedpage/webhook of this site","webhook URLは、このサイトの/ komoju / hostedpage / webhookにあります"
 "Payment Methods","お支払い方法"
-"Allow credit card payments","クレジットカード決済を許可する"
-"Allow konbini payments","コンビニ支払いを許可する"
 "Checkout Settings","決済設定"
 Title,題名
+"This plugin can only work if the store's base currency is one of: JPY","This plugin can only work if the store's base currency is one of: JPY"
```
Comment (Yuuki77) [inline src/app/code/Komoju/Payments/i18n/ja_JP.csv]:
```suggestion
"This plugin can only work if the store's base currency is one of: JPY","このプラグインは、ストアの基本通貨が日本円の場合にのみ動作します。"
```

File: src/app/code/Komoju/Payments/i18n/ja_JP.csv [hunk 5]
```diff
@@ -22,7 +22,9 @@ KOMOJU,KOMOJU
...
+"Encountered an issue communicating with KOMOJU. Please wait a moment and try again.","KOMOJUを読み込みませんでした。しばらく待ってから再実行してください。"
```
Comment (Yuuki77) [inline src/app/code/Komoju/Payments/i18n/ja_JP.csv]:
```suggestion
"Encountered an issue communicating with KOMOJU. Please wait a moment and try again.","KOMOJUとの通信に問題が発生しました。しばらく待ってから再実行してください。"
```

**Review by Yuuki77 — COMMENTED** [review]
Thank you for the screenshots! 🙏 
I made a few suggestions!
I think you can use "する" or without, but I thought it would be better without in this context.

**Comment by reneos** [comment]
@Yuuki77 Thank you very much for your help! 🙏🏻

**Comment by thoiberg** [comment]
> @thoiberg Hey Tim, added in the toggle ([here](https://github.com/degica/komoju-magento/pull/9/commits/3e3bdf495fbe0982cf841826ffa157750928358c)) for showing the title above the payment methods. It's not displayed by default.

Awesome thanks @reneos! I had one question in regards to the code, where are we actually figuring out what the title should be? I'm assuming we're generating it based on what payment methods the merchant has available, but I couldn't see where that code was in the change, and wanted to confirm.

**Comment by reneos** [comment]
@thoiberg Ah yeah, the title comes from the field just above the toggle in my screenshot (Title). The field for title is [here](https://github.com/degica/komoju-magento/blob/fa682f2789c8c0f60f664805b4edb7482e351e39/src/app/code/Komoju/Payments/etc/adminhtml/system.xml#L40-L43) in system.xml. Then the default can be set in config.xml [here](https://github.com/degica/komoju-magento/blob/fa682f2789c8c0f60f664805b4edb7482e351e39/src/app/code/Komoju/Payments/etc/config.xml#L11). These were already there when I started 😄 . I think the same title also controls what the payment method is called on the admin dashboard when it shows how customers paid.

**Review by thoiberg — APPROVED** [review]
LGTM ✨

**Comment by thoiberg** [comment]
@reneos great job with the magento plugin 🎉🎉 ! I'm not sure if we're planning to release this as a version or wait for more updates, but if we do let me know and I can walk you through the magento upload process (or what I still remember of it 😄 )

**Pull request merged**